pubnub-hyper/src/transport/hyper/pam.rs

```diff
@@ -0,0 +1,275 @@
+        data["spaces"] = map(&input.spaces);
+        data
+    };
+    let patterns = {
+        let input = &input.permissions.patterns;
+        let mut data = json::JsonValue::new_object();
+        data["channels"] = map(&input.channels);
+        data["groups"] = map(&input.groups);
+        data["users"] = map(&input.users);
+        data["spaces"] = map(&input.spaces);
+        data
+    };
+    let permissions = {
+        let mut data = json::JsonValue::new_object();
+        data["resources"] = resources;
+        data["patterns"] = patterns;
+        data["meta"] = input.permissions.meta;
+        data
+    };
+    let grant_body = {
+        let mut data = json::JsonValue::new_object();
+        data["ttl"] = input.ttl.into();
+        data["permissions"] = permissions;
+        data
+    };
+    json::stringify(grant_body)
+}
+
+/// Obtain UNIX timestamp.
+fn get_unix_time() -> u64 {
+    let current = std::time::SystemTime::now();
+    let since_the_epoch = current
+        .duration_since(std::time::UNIX_EPOCH)
+        .expect("things seem to be happening before the unix epoch, check system clock");
+    since_the_epoch.as_secs()
+}
+
+/// Prepare the signature.
+fn prepare_signature(
+    secret_key: &str,
+    subscribe_key: &str,
+    publish_key: &str,
+    timestamp: u64,
+    body: &str,
+) -> String {
+    pam_signature::sign(
+        secret_key,
+        pam_signature::Request {
+            publish_key,
+            method: "POST",
+            path: &format!("/v3/pam/{}/grant", subscribe_key),
+            query: &format!("timestamp={}", timestamp),
+            body,
+        },
+    )
+}
+
+async fn handle_grant_response(response: Response<Body>) -> Result<response::Grant, error::Error> {
+    match response.status() {
+        StatusCode::OK => {
+            let data_json = handle_json_response(response).await?;
+            let err_fn = || error::Error::UnexpectedResponseSchema(data_json.clone());
+            let token = {
+                let data = json_as_object(&data_json["data"]).ok_or_else(err_fn)?;
+                let token = data["token"].as_str().ok_or_else(err_fn)?;
+                token.to_owned()
+            };
+            Ok(token)
+        }
+        StatusCode::BAD_REQUEST | StatusCode::FORBIDDEN => {
+            let data = handle_json_response(response).await?;
+            let error_message: String = format!("{}", data["error"]["message"]);
+            Err(error::Error::Server(error_message))
+        }
+        _ => Err(error::Error::Server(format!(
+            "Server responded with an unexpected status code: {}",
+            response.status()
+        ))),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::pam;
+    use super::prepare_grant_body;
+    use std::collections::HashMap;
+
+    // The test data is huge, but the test logic is simple. We want to test
+    // every field, so we disable `clippy::too_many_lines` here.
+    #[allow(clippy::too_many_lines)]
```

Comment:
I used this approach where we had JSON outputs, but here I don't want to involve and additional complexity layers since this is a parser test. I.e. using JSON here would increase the uncertainty of where the issue comes from if the test starts failing.

A workaround would be to split the files into `.rs` includes here, but I think that'd be even worse.